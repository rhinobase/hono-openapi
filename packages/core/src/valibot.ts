import { type Hook, vValidator } from "@hono/valibot-validator";
import { type ConversionConfig, toJsonSchema } from "@valibot/to-json-schema";
import type {
  Env,
  Input as HonoInput,
  MiddlewareHandler,
  ValidationTargets,
} from "hono";
import {
  type BaseIssue,
  type BaseSchema,
  type GenericSchema,
  type GenericSchemaAsync,
  type InferInput,
  type InferOutput,
  parseAsync,
} from "valibot";
import convert from "./toOpenAPISchema.js";
import type {
  HasUndefined,
  OpenAPIRouteHandlerConfig,
  ResolverResult,
} from "./types.js";
import { generateValidatorDocs, uniqueSymbol } from "./utils.js";

/**
 * Generate a resolver for a Valibot schema
 * @param schema Valibot schema
 * @returns Resolver result
 */
export function resolver<
  T extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
>(schema: T, config?: ConversionConfig): ResolverResult {
  return {
    builder: async () => ({
      schema: await convert(toJsonSchema(schema, config)),
    }),
    validator: async (value) => {
      await parseAsync(schema, value);
    },
  };
}

/**
 * Create a validator middleware
 * @param target Target for validation
 * @param schema Valibot schema
 * @param hook Hook for validation
 * @returns Middleware handler
 */
export function validator<
  T extends GenericSchema | GenericSchemaAsync,
  Target extends keyof ValidationTargets,
  E extends Env,
  P extends string,
  In = InferInput<T>,
  Out = InferOutput<T>,
  I extends HonoInput = {
    in: HasUndefined<In> extends true
      ? {
          [K in Target]?: K extends "json"
            ? In
            : HasUndefined<keyof ValidationTargets[K]> extends true
              ? { [K2 in keyof In]?: ValidationTargets[K][K2] }
              : { [K2 in keyof In]: ValidationTargets[K][K2] };
        }
      : {
          [K in Target]: K extends "json"
            ? In
            : HasUndefined<keyof ValidationTargets[K]> extends true
              ? { [K2 in keyof In]?: ValidationTargets[K][K2] }
              : { [K2 in keyof In]: ValidationTargets[K][K2] };
        };
    out: { [K in Target]: Out };
  },
  V extends I = I,
>(target: Target, schema: T, hook?: Hook<T, E, P>): MiddlewareHandler<E, P, V> {
  const middleware = vValidator(target, schema, hook);

  // @ts-expect-error not typed well
  return Object.assign(middleware, {
    [uniqueSymbol]: {
      resolver: async (config: OpenAPIRouteHandlerConfig) =>
        generateValidatorDocs(
          target,
          // @ts-expect-error Need to fix the type
          await resolver(schema, { errorMode: "ignore" }).builder(config),
        ),
    },
  });
}
