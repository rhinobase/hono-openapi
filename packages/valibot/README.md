# valibot

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build valibot` to build the library.
