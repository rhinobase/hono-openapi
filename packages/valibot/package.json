{
  "name": "valibot-openapi",
  "version": "0.0.1",
  "dependencies": {
    "@valibot/to-json-schema": "^1.0.0-beta.2",
    "openapi-json-schema": "^2.0.0"
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.js",
  "files": [
    "dist",
    "!**/*.tsbuildinfo"
  ],
  "peerDependencies": {
    "valibot": "^1.0.0-beta.3",
    "openapi-types": "^12.1.3"
  },
  "nx": {
    "name": "valibot"
  }
}