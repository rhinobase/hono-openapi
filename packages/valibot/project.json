{
  "name": "valibot",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/valibot/src",
  "projectType": "library",
  "tags": [],
  "// targets": "to see all targets run: nx show project valibot --web",
  "targets": {
    "nx-release-publish": {
      "options": {
        "packageRoot": "{projectRoot}/dist"
      }
    }
  }
}
