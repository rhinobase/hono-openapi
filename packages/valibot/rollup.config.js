const { withNx } = require("@nx/rollup/with-nx");
const terser = require("@rollup/plugin-terser");

module.exports = withNx(
  {
    main: "./src/index.ts",
    outputPath: "./dist",
    tsConfig: "./tsconfig.lib.json",
    compiler: "swc",
    format: ["cjs", "esm"],
    assets: [{ input: ".", output: ".", glob: "*.md" }],
  },
  {
    plugins: [terser()],
  },
);
