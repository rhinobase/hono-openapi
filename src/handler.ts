import type { Context, Env, Hono } from "hono";
import type {
  BlankEnv,
  BlankInput,
  BlankSchema,
  Input,
  Schema,
} from "hono/types";
import type { OpenAPIV3_1 } from "openapi-types";
import type {
  DescribeRouteOptions,
  GenerateSpecOptions,
  ResolverReturnType,
  SanitizedGenerateSpecOptions,
} from "./types";
import {
  ALLOWED_METHODS,
  type AllowedMethods,
  registerSchemaPath,
  removeExcludedPaths,
  uniqueSymbol,
} from "./utils";

const DEFAULT_OPTIONS: Partial<GenerateSpecOptions> = {
  documentation: {},
  excludeStaticFile: true,
  exclude: [],
  excludeMethods: ["OPTIONS"],
  excludeTags: [],
};

type SpecContext = {
  components: Record<string, OpenAPIV3_1.SchemaObject>;
  options: SanitizedGenerateSpecOptions;
};

/**
 * Generate OpenAPI specs for the given Hono instance
 * @param hono Instance of Hono
 * @param options Options for generating OpenAPI specs
 * @param config Configuration for OpenAPI route handler
 * @param Context Route context for hiding routes
 * @returns OpenAPI specs
 */
export async function generateSpecs<
  E extends Env = BlankEnv,
  P extends string = string,
  I extends Input = BlankInput,
  S extends Schema = BlankSchema,
>(
  hono: Hono<E, S, P>,
  options = DEFAULT_OPTIONS,
  c?: Context<E, P, I>,
) {
  const ctx: SpecContext = {
    components: {},
    // @ts-expect-error
    options: {
      ...DEFAULT_OPTIONS,
      ...options,
    },
  };

  const _documentation = ctx.options.documentation ?? {};
  const schema = await generatePaths(hono, ctx);

  // Hide routes
  for (const path in schema) {
    for (const method in schema[path]) {
      const isHidden = getHiddenValue({
        valueOrFunc: schema[path][method]?.hide,
        method,
        path,
        c,
      });

      if (isHidden) {
        delete schema[path][method];
      }
    }
  }

  return {
    openapi: "3.1.0",
    ..._documentation,
    tags: _documentation.tags?.filter(
      (tag) => !ctx.options.excludeTags?.includes(tag?.name),
    ),
    info: {
      title: "Hono Documentation",
      description: "Development documentation",
      version: "0.0.0",
      ..._documentation.info,
    },
    paths: {
      ...removeExcludedPaths(schema, ctx),
      ..._documentation.paths,
    },
    components: {
      ..._documentation.components,
      schemas: {
        ...ctx.components,
        ..._documentation.components?.schemas,
      },
    },
  } satisfies OpenAPIV3_1.Document;
}

async function generatePaths<
  E extends Env = BlankEnv,
  P extends string = string,
  S extends Schema = BlankSchema,
>(
  hono: Hono<E, S, P>,
  ctx: SpecContext,
): Promise<OpenAPIV3_1.PathsObject> {
  const paths: OpenAPIV3_1.PathsObject = {};

  for (const route of hono.routes) {
    // Finding routes with uniqueSymbol
    if (!(uniqueSymbol in route.handler)) {
      // Include empty paths, if enabled
      if (ctx.options.includeEmptyPaths) {
        registerSchemaPath({
          route,
          paths,
        });
      }

      continue;
    }

    const routeMethod = route.method as AllowedMethods | "ALL";

    // All method acts like a middleware, so we can skip it
    if (routeMethod !== "ALL") {
      // Exclude methods
      if (ctx.options.excludeMethods?.includes(routeMethod)) {
        continue;
      }

      // Include only allowed methods
      if (!ALLOWED_METHODS.includes(routeMethod)) {
        continue;
      }
    }

    const middlewareHandler = route.handler[uniqueSymbol] as
      | ResolverReturnType
      | { schema: DescribeRouteOptions; components: undefined };

    const defaultOptionsForThisMethod = ctx.options.defaultOptions
      ?.[routeMethod];

    const { schema: routeSpecs, components = {} } =
      "schema" in middlewareHandler
        ? middlewareHandler
        : await middlewareHandler
          .toOpenAPISchema({
            ...defaultOptionsForThisMethod,
            components: ctx.components,
          });

    ctx.components = {
      ...ctx.components,
      ...components,
    };

    registerSchemaPath({
      route,
      specs: routeSpecs,
      paths,
    });
  }

  return paths;
}

function getHiddenValue(options: {
  valueOrFunc: DescribeRouteOptions["hide"];
  c?: Context;
  method: string;
  path: string;
}) {
  const { valueOrFunc, c, method, path } = options;

  if (valueOrFunc != null) {
    if (typeof valueOrFunc === "boolean") {
      return valueOrFunc;
    } else if (typeof valueOrFunc === "function") {
      if (c) {
        return valueOrFunc(c);
      } else {
        console.warn(
          `'c' is not defined, cannot evaluate hide function for ${method} ${path}`,
        );
      }
    }
  }

  return false;
}
