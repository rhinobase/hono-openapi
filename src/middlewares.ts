import { type Hook, sValidator } from "@hono/standard-validator";
import type { Env, Input, MiddlewareHandler, ValidationTargets } from "hono";
import type { StandardSchemaV1 } from "@standard-schema/spec";
import { toJsonSchema } from "@standard-community/standard-json";
import { toOpenAPISchema } from "@standard-community/standard-openapi";
import { uniqueSymbol } from "./utils";
import { DescribeRouteOptions } from "./types";

/**
 * Generate a resolver for a validation schema
 * @param schema Validation schema
 * @returns Resolver result
 */
export function resolver<Schema extends StandardSchemaV1>(schema: Schema) {
  return {
    vendor: schema["~standard"].vendor,
    validate: schema["~standard"].validate,
    toJSONSchema: (options?: Record<string, unknown>) =>
      toJsonSchema(schema, options),
    toOpenAPI: (options?: Record<string, unknown>) =>
      toOpenAPISchema(schema, options),
  };
}

type HasUndefined<T> = undefined extends T ? true : false;

/**
 * Create a validator middleware
 * @param target Target for validation
 * @param schema Validation schema
 * @param hook Hook for validation
 * @returns Middleware handler
 */
export function validator<
  Schema extends StandardSchemaV1,
  Target extends keyof ValidationTargets,
  E extends Env,
  P extends string,
  In = StandardSchemaV1.InferInput<Schema>,
  Out = StandardSchemaV1.InferOutput<Schema>,
  I extends Input = {
    in: HasUndefined<In> extends true ? {
        [K in Target]?: In extends ValidationTargets[K] ? In
          : { [K2 in keyof In]?: ValidationTargets[K][K2] };
      }
      : {
        [K in Target]: In extends ValidationTargets[K] ? In
          : { [K2 in keyof In]: ValidationTargets[K][K2] };
      };
    out: { [K in Target]: Out };
  },
  V extends I = I,
>(
  target: Target,
  schema: Schema,
  hook?: Hook<StandardSchemaV1.InferOutput<Schema>, E, P, Target>,
  options?: Record<string, unknown>,
): MiddlewareHandler<E, P, V> {
  const middleware = sValidator(target, schema, hook);

  // @ts-expect-error not typed well
  return Object.assign(middleware, {
    [uniqueSymbol]: {
      target,
      ...resolver(schema),
      options,
    },
  });
}

/**
 * Describe a route with OpenAPI specs.
 * @param specs Options for describing a route
 * @returns Middleware handler
 */
export function describeRoute(specs: DescribeRouteOptions): MiddlewareHandler {
  const middleware: MiddlewareHandler = async (_c, next) => {
    await next();
  };

  return Object.assign(middleware, {
    [uniqueSymbol]: {
      specs,
    },
  });
}
