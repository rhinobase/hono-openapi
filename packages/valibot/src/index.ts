import type { GenericSchema, GenericSchemaAsync } from "valibot";
import * as v from "valibot";
import type {
  ConversionContext,
  ConversionConfig,
  ConversionResponse,
} from "./types";
import { convertSchema } from "./convertSchema";

export type OpenAPIMetadata<T = string> = {
  description: string;
  example: T;
  examples: [T, ...T[]];
  default: T;
  ref: string;
};

export const metadata = <
  TInput,
  TMetadata extends Partial<OpenAPIMetadata<TInput>>
>(
  metadata: TMetadata
) => v.metadata<TInput, TMetadata>(metadata);

/**
 * Converts a Valibot schema to the OpenAPI Schema format.
 *
 * @param schema The Valibot schema object.
 * @param config The OpenAPI Schema configuration.
 *
 * @returns The converted OpenAPI Schema.
 */
export function createSchema<T extends GenericSchema | GenericSchemaAsync>(
  schema: T,
  config?: ConversionConfig
): ConversionResponse {
  // Initialize JSON Schema context
  const context: ConversionContext = {
    definitions: {},
    referenceMap: new Map(),
    getterMap: new Map(),
  };

  const openAPISchema = convertSchema({}, schema, config, context);

  return openAPISchema;
}
