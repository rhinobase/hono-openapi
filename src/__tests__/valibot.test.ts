import { Hono } from "hono";
import * as v from "valibot";
import { describe, expect, it } from "vitest";
import { generateSpecs } from "../handler.js";
import { describeRoute, resolver, validator } from "../middlewares.js";

describe("valibot", () => {
  it("basic", async () => {
    const app = new Hono().get(
      "/",
      describeRoute({
        tags: ["test"],
        summary: "Test route",
        description: "This is a test route",
        responses: {
          200: {
            description: "Success",
            content: {
              "application/json": {
                schema: resolver(
                  v.object({
                    message: v.string(),
                  }),
                ),
              },
            },
          },
        },
      }),
      validator("json", v.object({ message: v.string() })),
      async (c) => {
        return c.json({ message: "Hello, world!" });
      },
    );

    const specs = await generateSpecs(app);

    expect(specs).toMatchSnapshot();
  });

  it("with metadata", async () => {
    const app = new Hono().get(
      "/",
      describeRoute({
        tags: ["test"],
        summary: "Test route",
        description: "This is a test route",
        responses: {
          200: {
            description: "Success",
            content: {
              "application/json": {
                schema: resolver(
                  v.pipe(
                    v.object({
                      message: v.string(),
                    }),
                    v.metadata({ ref: "SuccessResponse" }),
                  ),
                ),
              },
            },
          },
        },
      }),
      validator("json", v.object({ message: v.string() })),
      async (c) => {
        return c.json({ message: "Hello, world!" });
      },
    );

    const specs = await generateSpecs(app);

    expect(specs).toMatchSnapshot();
  });

  it("with transformation", async () => {
    const app = new Hono().get(
      "/",
      describeRoute({
        tags: ["test"],
        summary: "Test route",
        description: "This is a test route",
        responses: {
          200: {
            description: "Success",
            content: {
              "application/json": {
                schema: resolver(v.string()),
              },
            },
          },
        },
      }),
      validator(
        "query",
        v.object({
          names: v.pipe(
            v.string(),
            v.transform((val) => val.split("|")),
            v.array(v.string()),
          ),
        }),
        undefined,
        {
          options: { typeMode: "output" },
        },
      ),
      async (c) => {
        const { names } = await c.req.valid("query");
        return c.json({ message: `Hello ${names.join(", ")}!` });
      },
    );

    const specs = await generateSpecs(app);

    expect(specs).toMatchSnapshot();
  });
});
