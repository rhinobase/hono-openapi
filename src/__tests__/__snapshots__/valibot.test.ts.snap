// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`valibot > basic 1`] = `
{
  "components": {
    "schemas": {},
  },
  "info": {
    "description": "Development documentation",
    "title": "Hono Documentation",
    "version": "0.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "get": {
        "description": "This is a test route",
        "operationId": "GETIndex",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "required": [
                  "message",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$schema": "http://json-schema.org/draft-07/schema#",
                  "properties": {
                    "message": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "message",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
        "summary": "Test route",
        "tags": [
          "test",
        ],
      },
    },
  },
  "tags": undefined,
}
`;

exports[`valibot > with metadata 1`] = `
{
  "components": {
    "schemas": {
      "SuccessResponse": {
        "properties": {
          "message": {
            "type": "string",
          },
        },
        "required": [
          "message",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "description": "Development documentation",
    "title": "Hono Documentation",
    "version": "0.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/": {
      "get": {
        "description": "This is a test route",
        "operationId": "GETIndex",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "properties": {
                  "message": {
                    "type": "string",
                  },
                },
                "required": [
                  "message",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/SuccessResponse",
                },
              },
            },
            "description": "Success",
          },
        },
        "summary": "Test route",
        "tags": [
          "test",
        ],
      },
    },
  },
  "tags": undefined,
}
`;
